File: kernel/src/main/kotlin/Rpc.kt
package com.kjrandez.context.kernel

import com.kjrandez.context.kernel.entity.*
import kotlinx.serialization.PolymorphicSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.coroutines.Continuation
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

interface RpcMessage
interface RpcArgument
interface RpcDataClass

@Serializable @SerialName("int")
data class RpcInt(val value: Int) : RpcArgument

@Serializable @SerialName("float")
data class RpcFloat(val value: Double) : RpcArgument

@Serializable @SerialName("string")
data class RpcString(val value: String) : RpcArgument

@Serializable @SerialName("bool")
data class RpcBool(val value: Boolean) : RpcArgument

@Serializable @SerialName("list")
data class RpcList(val value: List<RpcArgument?>) : RpcArgument

@Serializable @SerialName("dict")
data class RpcDict(val value: Map<String, RpcArgument?>) : RpcArgument

@Serializable @SerialName("hostObj")
data class RpcHostObj(val value: Int) : RpcArgument

@Serializable @SerialName("clientObj")
data class RpcClientObj(val value: Int) : RpcArgument

@Serializable @SerialName("data")
data class RpcData(val value: RpcDataClass) : RpcArgument

@Serializable @SerialName("send")
data class RpcSend (
    val target: Int,
    val selector: String,
    val arguments: List<RpcArgument?>
) : RpcMessage

@Serializable @SerialName("call")
data class RpcCall (
    val id: Int,
    val target: Int,
    val selector: String,
    val arguments: List<RpcArgument?>
) : RpcMessage

@Serializable @SerialName("yield")
data class RpcYield (
    val id: Int,
    val result: RpcArgument?
) : RpcMessage

@Serializable @SerialName("error")
data class RpcError (
    val id : Int?,
    val message: String
) : RpcMessage

class RpcEncodeException(message: String) : Exception(message)

class Proxy (
    val eid: Int,
    val call: suspend (selector: String, args: List<Any?>) -> Any?,
    val send: (selector: String, args: List<Any?>) -> Unit
)

class Rpc(private val database: Database, private val hostService: Entity, private val sendMessage: (String) -> Unit)
{
    private val json = rpcJson(database)

    private val proxyMap = mutableMapOf<Int, Proxy>()
    private val callMap = mutableMapOf<Int, Continuation<Any?>>()
    private var nextCallId = 0

    val clientService = resolveProxy(0)

    suspend fun receive(message: String) {
        val desc = parse(message)
        when (desc) {
            is RpcCall -> dispatchCall(desc)
            is RpcSend -> dispatchSend(desc)
            is RpcYield -> dispatchYield(desc)
            is RpcError -> dispatchError(desc)
            else -> Unit
        }
    }

    private suspend fun dispatchCall(desc: RpcCall) {
        try {
            // Throws DatabaseException, EntityException, or RpcEncodingException
            val entity = resolveEntity(desc.target)
            val result = entity.invoke(desc.selector, desc.arguments.map { unwrapRpcArgument(it) }.toTypedArray())
            val yieldDesc = RpcYield(desc.id, wrapRpcArgument(result))
            sendMessage(stringify(yieldDesc))
        } catch(ex: Exception) {
            ex.printStackTrace()
            sendMessage(stringify(RpcError(desc.id, ex.message ?: "N/A")))
        }
    }

    private suspend fun dispatchSend(desc: RpcSend) {
        val entity = resolveEntity(desc.target)

        try {
            entity.invoke(desc.selector, desc.arguments.map { unwrapRpcArgument(it) }.toTypedArray())
        } catch(ex: Exception) {
            ex.printStackTrace()
            sendMessage(stringify(RpcError(null, ex.message ?: "N/A")))
        }
    }

    private fun dispatchYield(desc: RpcYield) {
        val cont = callMap[desc.id]
        if (cont != null) {
            callMap.remove(desc.id)
            cont.resume(unwrapRpcArgument(desc.result))
        }
    }

    private fun dispatchError(desc: RpcError) {
        if (desc.id != null) {
            val cont = callMap[desc.id]
            if (cont != null) {
                callMap.remove(desc.id)
                cont.resumeWithException(RpcEncodeException("Remote error: ${desc.message}"))
            }
        }
    }

    private suspend fun call(target: Int, selector: String, args: List<Any?>): Any? {
        return suspendCoroutine { cont ->
            val callId = nextCallId
            callMap[callId] = cont

            val callDesc = RpcCall(callId, target, selector, args.map { wrapRpcArgument(it) })
            sendMessage(stringify(callDesc))
        }
    }

    private fun send(target: Int, selector: String, args: List<Any?>) {
        val sendDesc = RpcSend(target, selector, args.map { wrapRpcArgument(it) })
        sendMessage(stringify(sendDesc))
    }

    private fun stringify(message: RpcMessage) = json.stringify(PolymorphicSerializer(RpcMessage::class), message)
    private fun parse(message: String) = json.parse(PolymorphicSerializer(RpcMessage::class), message)

    private fun wrapRpcArgument(arg: Any?): RpcArgument? {
        return when (arg) {
            is List<*> -> RpcList(arg.map { wrapRpcArgument(it) })
            is Map<*, *> -> RpcDict((arg as Map<String, *>).mapValues { wrapRpcArgument (it.value) })
            is Int -> RpcInt(arg)
            is Double -> RpcFloat(arg)
            is String -> RpcString(arg)
            is Boolean -> RpcBool(arg)
            is Entity -> RpcHostObj(arg.eid)
            is Proxy -> RpcClientObj(arg.eid)
            is RpcDataClass -> RpcData(arg)
            null -> null
            else -> throw RpcEncodeException("No encoding for argument type")
        }
    }

    private fun unwrapRpcArgument(arg: RpcArgument?): Any? {
        return when (arg) {
            is RpcList -> arg.value.map { unwrapRpcArgument(it) }
            is RpcDict -> arg.value.mapValues { unwrapRpcArgument(it.value) }
            is RpcInt -> arg.value
            is RpcFloat -> arg.value
            is RpcString -> arg.value
            is RpcBool -> arg.value
            is RpcHostObj -> resolveEntity(arg.value)
            is RpcClientObj -> resolveProxy(arg.value)
            is RpcData -> arg.value
            null -> null
            else -> Unit
        }
    }

    private fun resolveProxy(eid: Int): Proxy {
        return proxyMap[eid] ?: Proxy(
            eid,
            { selector, args -> this.call(eid, selector, args) },
            { selector, args -> this.send(eid, selector, args) }
        ).also { proxyMap[eid] = it }
    }

    private fun resolveEntity(eid: Int): Entity {
        return database.lookup(eid)
    }
}

File: pycontext/context/rpc.py
class Proxy:
    def __init__(self, foreignId, clientCall, clientSend, queue):
        self.id = foreignId
        self.queue = queue
        self.clientCall = clientCall
        self.clientSend = clientSend

    def callBlocking(self, selector, arguments):
        return self.queue.awaitResult(lambda: self.call(selector, arguments))

    def sendBlocking(self, selector, arguments):
        self.queue.awaitResult(lambda: self.send(selector, arguments))

    async def call(self, selector: str, arguments):
        return await self.clientCall(self.id, selector, arguments)

    async def send(self, selector: str, arguments):
        await self.clientSend(self.id, selector, arguments)


class RpcException(Exception):
    def __init__(self, message):
        super().__init__(message)


class Rpc:
    def __init__(self, loop, queue, clientService, send):
        self.loop = loop
        self.queue = queue
        self.clientService = clientService
        self.send = send
        self.proxyMap = {}
        self.callMap = {}
        self.nextCallId = 0

    async def clientCall(self, targetId, selector, arguments):
        argDescs = [self.wrapArgument(X) for X in arguments]

        callId, future = self.nextFuture()

        await self.send({
            'type': 'call',
            'id': callId,
            'target': targetId,
            'selector': selector,
            'arguments': argDescs
        })

        try:
            result = await future
            return result
        except Exception as ex:
            print("Exception waiting for call result future")
            raise ex

    async def clientSend(self, targetId, selector, arguments):
        argDescs = [self.wrapArgument(X) for X in arguments]

        await self.send({
            'type': 'send',
            'target': targetId,
            'selector': selector,
            'arguments': argDescs
        })

    async def receive(self, msg):
        if msg['type'] == 'send':
            self.dispatchSend(msg['target'], msg['selector'], msg['arguments'])
        elif msg['type'] == 'call':
            await self.dispatchCall(msg['id'], msg['target'], msg['selector'], msg['arguments'])
        elif msg['type'] == 'yield':
            self.dispatchYield(msg['id'], msg['result'])
        elif msg['type'] == 'error':
            self.dispatchError(msg['id'], msg['message'])
        else:
            print("Unhandled message")

    async def dispatchCall(self, foreignId, targetId, selector: str, argDescs):
        target = self.clientService
        if targetId != 0: # expand to support whole dictionary of client objects
            target = self.dataset.lookup(targetId)

        method = getattr(target, selector)
        arguments = map(lambda X: self.unwrapArgument(X), argDescs)

        result = method(*arguments)

        await self.send({
            'type': 'return',
            'id': foreignId,
            'result': self.wrapArgument(result)
        })

    def dispatchSend(self, targetId, selector, argDescs):
        target = self.clientService
        if targetId != 0:
            target = self.dataset.lookup(targetId)

        method = getattr(target, selector)
        arguments = map(lambda X: self.unwrapArgument(X), argDescs)

        method(*arguments)

    def dispatchYield(self, localId, resultDesc):
        future = self.callMap[localId]
        del self.callMap[localId]

        future.set_result(self.unwrapArgument(resultDesc))

    def dispatchError(self, localId, message):
        if localId is not None:
            future = self.callMap[localId]
            del self.callMap[localId]

            future.set_exception(RpcException("Remote Exception: " + message))

    def inflateDataObject(self, obj):
        if isinstance(obj, list):
            return [self.inflateDataObject(element) for element in obj]
        elif isinstance(obj, dict):
            if 'type' in obj and obj['type'] == "hostObj":
                return self.resolveProxy(obj['value'])
            elif 'type' in obj and obj['type'] == "clientObj":
                return self.dataset.lookup(obj['value'])
            else:
                return {K: self.inflateDataObject(V) for K, V in obj.items()}
        else:
            return obj

    def unwrapArgument(self, arg):
        argType, argValue = arg['type'], arg['value']

        if argType == 'clientObj':
            return self.dataset.lookup(argValue)
        elif argType == 'hostObj':
            return self.resolveProxy(argValue)
        elif argType == 'list':
            return [self.unwrapArgument(X) for X in argValue]
        elif argType == 'map':
            return {K: self.unwrapArgumennt(V) for K, V in argValue.items()}
        elif argType == 'data':
            return self.inflateDataObject(arg['value'])
        else: # int, float, string, boolean decoded natively
            return arg['value']

    def wrapArgument(self, arg):
        if arg == self.clientService: # expand to support whole dictionary of client objects
            return {'type': 'clientObj', 'value': 0}
        elif isinstance(arg, Proxy):
            return {'type': 'hostObj', 'value': arg.id}
        elif isinstance(arg, list):
            return {
                'type': 'list',
                'value': [self.wrapArgument(X) for X in arg]
            }
        elif isinstance(arg, dict):
            return {
                'type': 'map',
                'value': {K: self.wrapArgument(V) for K, V in arg}
            }
        elif isinstance(arg, int):
            return {'type': 'int', 'value': arg}
        elif isinstance(arg, float):
            return {'type': 'float', 'value': arg}
        elif isinstance(arg, bool):
            return {'type': 'bool', 'value': arg}
        elif isinstance(arg, str):
            return {'type': 'string', 'value': arg}
        else:
            raise self.RpcException("No encoding for argument type")

    def resolveProxy(self, id: int):
        if id in self.proxyMap:
            return self.proxyMap[id]

        proxy = Proxy(id, self.clientCall, self.clientSend, self.queue)
        self.proxyMap[id] = proxy
        return proxy

    def nextFuture(self):
        callId = self.nextCallId
        future = self.loop.create_future()
        self.callMap[callId] = future
        self.nextCallId += 1

        return callId, future
